### PR TITLE
Ddp discrete

### PR DESCRIPTION
**Pull request by Sobczakus1658**
DDP discrete

### CODE REVIEW

File: training/loss.py [hunk 1]
```diff
@@ -80,3 +80,56 @@ def __call__(self, net, images, labels=None, augment_pipe=None):
...
+            self.alphas = net_model.alphas
+            self.sigmas = net_model.sigmas
+
+    def __call__(self, net, images, labels=None, augment_pipe=None):
+        if self.alphas is None:
+            self.setup_schedules(net)
+
+        # Mỏj loss jest w postaci (16), więc muszę wylosować t i obliczyć ε_t
+        # Generowanie z rozkładu jednostajnego
+        t = torch.randint(1, len(self.alphas), (images.shape[0],), device=images.device)
```
Comment (SzymonM213) [inline training/loss.py]: A my nie chcemy mieć t ~ U([0, 1])? W sensie robić trening tak jak edm że losujemy moment t z jednostajnego i liczymy loss dla tego jednego

File: training/loss.py [hunk 2]
```diff
@@ -80,3 +80,56 @@ def __call__(self, net, images, labels=None, augment_pipe=None):
...
+        # chwilowo wstawienie U_t jako 1 jako stała
+        gamma_s = (alpha_s/sigma_s)**2 * (sigma_t/alpha_t)**2
+
+        eta_s = (gamma_s - 1)/(torch.sqrt(gamma_s) + torch.sqrt(gamma_s - 1))
+        eta_s = eta_s.unsqueeze(1).unsqueeze(2).unsqueeze(3)
+        eta_s = eta_s.view(-1,1,1,1)
+
+        # Generowanie szumu
+        eps_t = torch.randn_like(images)
+        eps_s = torch.randn_like(images)
```
Comment (SzymonM213) [inline training/loss.py]: Mi się wydaje, że my nie musimy w treningu patrzeć na te 2 epsilony, po prostu trenujemy sieć do przewidywania U_t * eps_t, ale mogę halucynować też

File: training/loss.py [hunk 3]
```diff
@@ -80,3 +80,56 @@ def __call__(self, net, images, labels=None, augment_pipe=None):
...
+        z_t = alpha_t.view(-1,1,1,1)*images + sigma_t.view(-1,1,1,1)*epsilon_t
+        z_s = alpha_s.view(-1,1,1,1)*images + sigma_s.view(-1,1,1,1)*epsilon_s
+
+        # Przewidywanie ε_t, ze wzoru (2)
+        with torch.no_grad():
+            u_target = (z_s - alpha_s.view(-1,1,1,1)*images)/(sigma_s.view(-1,1,1,1))
+        u_pred = net(z_t, sigma_t, labels)
+
+        # wzór 16
+        loss = 0.5 * ((u_target - u_pred)**2).mean()
```
Comment (SzymonM213) [inline training/loss.py]: 1/2 niepotrzebna bo stała w lossie nie ma znaczenia ale to szczegół

File: training/networks.py [hunk 4]
```diff
@@ -671,3 +671,80 @@ def round_sigma(self, sigma):
...
+            min_ratio = (torch.sqrt(gamma - 1) / self.Ut.abs()).max().clamp(min=1.0)
+            self.Ut.data *= min_ratio
+            # wzór z końca kartki
+            numerator = gamma - 1
+            denominator = torch.sqrt(gamma)*self.Ut**2 + torch.sqrt(self.Ut**2 + 1 - gamma)
+            etas = numerator / denominator
+            self.register_buffer('etas', etas)
+
+    def forward(self, x, sigma, class_labels=None, force_fp32=False, **model_kwargs):
+        # parametry z EDM - nie wiem czy finalnie będziemy to ruszać, może to do usunięcia
```
Comment (SzymonM213) [inline training/networks.py]: do usunięcia

File: training/networks.py [hunk 5]
```diff
@@ -671,3 +671,80 @@ def round_sigma(self, sigma):
...
+
+        t_idx = self.sigma_to_t(sigma)
+        alpha_t = self.alphas[t_idx].reshape(-1, 1, 1, 1)
+        sigma_t = self.sigmas[t_idx].reshape(-1, 1, 1, 1)
+        eta_t = self.etas[t_idx].reshape(-1, 1, 1, 1)
+
+        # wzór (1)
+        eps_t = torch.randn_like(x)
+        eps_t_prime = torch.randn_like(x)
+        zs = alpha_t * x + sigma_t * torch.sqrt(1 - eta_t**2) * eps_t + sigma_t * eta_t * eps_t_prime
```
Comment (SzymonM213) [inline training/networks.py]: A x to nie jest już zaszumiony obraz? Trochę bez sensu żeby forward w modelu przyjmowało już to co jest celem

File: training/networks.py [hunk 6]
```diff
@@ -671,3 +671,80 @@ def round_sigma(self, sigma):
...
+
+        F_x = self.model(zs.to(x.dtype), c_noise.flatten(), class_labels=class_labels, **model_kwargs)
+
+        # ro również może do wywalenia, wzorowane na EDM
+        D_x = c_skip * x + c_out * F_x.to(torch.float32)
```
Comment (SzymonM213) [inline training/networks.py]: Wydaje mi się, że tego nie powinno być, bo edm sieć do przewidywania 1/c_out * (y - c_skip * (y + noise)) i stąd te stałe, jak my chcemy po prostu przewidywać U_t * noise to pewnie najładniej byłoby wrzucić zamiast tego co tu jest 
D_x = (x - F_x) * U_t  (czyli eps_theta * U_t)
i w loss.py zwracać net(images) - eps_t * U_t

File: training/loss.py (hunk 1 above)
Comment (Sobczakus1658) [reply training/loss.py]: W network.py tak inicjalizuje sobie alphy:

        t = torch.linspace(0, 1, num_timesteps)
        alphas = torch.cos(t * (torch.pi / 2))
        self.register_buffer('alphas', alphas)
        
 indeksy więc są od 0 do num_timesteps -1 i są liczbami całkowitymi.

File: training/loss.py (hunk 3 above)
Comment (Sobczakus1658) [reply training/loss.py]: zgoda, ale chyba zostawię, bo tak jest we wzorze od niego.

File: training/loss.py (hunk 2 above)
Comment (Sobczakus1658) [reply training/loss.py]: Chyba zgodzę się, że takie rozwiązanie jest niepoprawne, ale to co proponujesz to chyba rozwiązanie dla eta = 0. Zastanawiam się czy nie powiniem być epsilon_t zdefiniowany w network.py zgodnie ze wzorem (1) i wtedy 

eps_t = self.eps[t]
eps_s = torch.randn_like(images)
